File: src/agents/bee/agent.ts
/**
 * Copyright 2024 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseAgent } from "@/agents/base.js";
import { AnyTool } from "@/tools/base.js";
import { BaseMemory } from "@/memory/base.js";
import { ChatLLM, ChatLLMOutput } from "@/llms/chat.js";
import { BaseMessage, Role } from "@/llms/primitives/message.js";
import { AgentMeta } from "@/agents/types.js";
import { Emitter } from "@/emitter/emitter.js";
import {
  BeeAgentTemplates,
  BeeCallbacks,
  BeeRunInput,
  BeeRunOptions,
  BeeRunOutput,
} from "@/agents/bee/types.js";
import { GetRunContext } from "@/context.js";
import { assign } from "@/internals/helpers/object.js";
import * as R from "remeda";
import { BaseRunner } from "@/agents/bee/runners/base.js";
import { GraniteRunner } from "@/agents/bee/runners/granite/runner.js";
import { DefaultRunner } from "@/agents/bee/runners/default/runner.js";
import { ValueError } from "@/errors.js";

export interface BeeInput {
  llm: ChatLLM<ChatLLMOutput>;
  tools: AnyTool[];
  memory: BaseMemory;
  meta?: Omit<AgentMeta, "tools">;
  templates?: Partial<BeeAgentTemplates>;
}

export class BeeAgent extends BaseAgent<BeeRunInput, BeeRunOutput, BeeRunOptions> {
  public readonly emitter = Emitter.root.child<BeeCallbacks>({
    namespace: ["agent", "bee"],
    creator: this,
  });

  protected runner: new (...args: ConstructorParameters<typeof BaseRunner>) => BaseRunner;

  constructor(protected readonly input: BeeInput) {
    super();

    const duplicate = input.tools.find((a, i, arr) =>
      arr.find((b, j) => i !== j && a.name.toUpperCase() === b.name.toUpperCase()),
    );
    if (duplicate) {
      throw new ValueError(
        `Agent's tools must all have different names. Conflicting tool: ${duplicate.name}.`,
      );
    }

    this.runner = this.input.llm.modelId.includes("granite") ? GraniteRunner : DefaultRunner;
  }

  static {
    this.register();
  }

  get memory() {
    return this.input.memory;
  }

  get meta(): AgentMeta {
    const tools = this.input.tools.slice();

    if (this.input.meta) {
      return { ...this.input.meta, tools };
    }

    return {
      name: "Bee",
      tools,
      description:
        "The Bee framework demonstrates its ability to auto-correct and adapt in real-time, improving the overall reliability and resilience of the system.",
      ...(tools.length > 0 && {
        extraDescription: [
          `Tools that I can use to accomplish given task.`,
          ...tools.map((tool) => `Tool '${tool.name}': ${tool.description}.`),
        ].join("\n"),
      }),
    };
  }

  protected async _run(
    input: BeeRunInput,
    options: BeeRunOptions = {},
    run: GetRunContext<typeof this>,
  ): Promise<BeeRunOutput> {
    const runner = new this.runner(
      this.input,
      {
        ...options,
        execution: options?.execution ?? {
          maxRetriesPerStep: 3,
          totalMaxRetries: 20,
          maxIterations: 10,
        },
      },
      run,
    );
    await runner.init(input);

    let finalMessage: BaseMessage | undefined;
    while (!finalMessage) {
      const { state, meta, emitter, signal } = await runner.createIteration();

      if (state.tool_name && state.tool_input) {
        const { output, success } = await runner.tool({
          state,
          emitter,
          meta,
          signal,
        });
        await runner.memory.add(
          BaseMessage.of({
            role: Role.ASSISTANT,
            text: runner.templates.assistant.render({
              thought: [state.thought].filter(R.isTruthy),
              toolName: [state.tool_name].filter(R.isTruthy),
              toolInput: [state.tool_input].filter(R.isTruthy).map((call) => JSON.stringify(call)),
              toolOutput: [output],
              finalAnswer: [state.final_answer].filter(R.isTruthy),
            }),
            meta: { success },
          }),
        );
        assign(state, { tool_output: output });

        for (const key of ["partialUpdate", "update"] as const) {
          await emitter.emit(key, {
            data: state,
            update: { key: "tool_output", value: output, parsedValue: output },
            meta: { success, ...meta },
            memory: runner.memory,
          });
        }
      }
      if (state.final_answer) {
        finalMessage = BaseMessage.of({
          role: Role.ASSISTANT,
          text: state.final_answer,
          meta: {
            createdAt: new Date(),
          },
        });
        await runner.memory.add(finalMessage);
        await emitter.emit("success", {
          data: finalMessage,
          iterations: runner.iterations,
          memory: runner.memory,
          meta,
        });
      }
    }

    if (input.prompt !== null) {
      await this.input.memory.add(
        BaseMessage.of({
          role: Role.USER,
          text: input.prompt,
          meta: {
            createdAt: run.createdAt,
          },
        }),
      );
    }

    await this.input.memory.add(finalMessage);
    return { result: finalMessage, iterations: runner.iterations, memory: runner.memory };
  }

  createSnapshot() {
    return {
      ...super.createSnapshot(),
      input: this.input,
      emitter: this.emitter,
      runner: this.runner,
    };
  }
}

File: python/beeai_framework/agents/bee/agent.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


from collections.abc import Callable
from datetime import UTC, datetime

from beeai_framework.agents.base import BaseAgent
from beeai_framework.agents.runners.base import (
    BaseRunner,
    BeeRunnerToolInput,
    BeeRunnerToolResult,
    RunnerIteration,
)
from beeai_framework.agents.runners.default.runner import DefaultRunner
from beeai_framework.agents.runners.granite.runner import GraniteRunner
from beeai_framework.agents.types import (
    AgentMeta,
    BeeAgentExecutionConfig,
    BeeInput,
    BeeRunInput,
    BeeRunOptions,
    BeeRunOutput,
)
from beeai_framework.backend import Message
from beeai_framework.backend.message import AssistantMessage, MessageMeta, UserMessage
from beeai_framework.context import RunContext
from beeai_framework.emitter import Emitter, EmitterInput
from beeai_framework.memory import BaseMemory


class BeeAgent(BaseAgent):
    runner: Callable[..., BaseRunner]

    def __init__(self, bee_input: BeeInput) -> None:
        self.input = bee_input
        if "granite" in self.input.llm.model_id:
            self.runner = GraniteRunner
        else:
            self.runner = DefaultRunner
        self.emitter = Emitter.root().child(
            EmitterInput(
                namespace=["agent", "bee"],
                creator=self,
            )
        )

    @property
    def memory(self) -> BaseMemory:
        return self.input.memory

    @memory.setter
    def memory(self, memory: BaseMemory) -> None:
        self.input.memory = memory

    @property
    def meta(self) -> AgentMeta:
        tools = self.input.tools[:]

        if self.input.meta:
            return AgentMeta(
                name=self.input.meta.name,
                description=self.input.meta.description,
                extra_description=self.input.meta.extra_description,
                tools=tools,
            )

        extra_description = ["Tools that I can use to accomplish given task."]
        for tool in tools:
            extra_description.append(f"Tool ${tool.name}': ${tool.description}.")

        return AgentMeta(
            name="BeeAI",
            tools=tools,
            description="The BeeAI framework demonstrates its ability to auto-correct and adapt in real-time, improving"
            " the overall reliability and resilience of the system.",
            extra_description="\n".join(extra_description) if len(tools) > 0 else None,
        )

    async def _run(self, run_input: BeeRunInput, options: BeeRunOptions | None, context: RunContext) -> BeeRunOutput:
        runner = self.runner(
            self.input,
            (
                options
                if options
                else BeeRunOptions(
                    execution=self.input.execution
                    or (options.execution if options is not None else None)
                    or BeeAgentExecutionConfig(
                        max_retries_per_step=3,
                        total_max_retries=20,
                        max_iterations=10,
                    ),
                    signal=None,
                )
            ),
            context,
        )
        await runner.init(run_input)

        final_message: Message | None = None
        while not final_message:
            iteration: RunnerIteration = await runner.create_iteration()

            if iteration.state.tool_name and iteration.state.tool_input:
                tool_result: BeeRunnerToolResult = await runner.tool(
                    input=BeeRunnerToolInput(
                        state=iteration.state,
                        emitter=iteration.emitter,
                        meta=iteration.meta,
                        signal=iteration.signal,
                    )
                )
                await runner.memory.add(
                    AssistantMessage(
                        content=runner.templates.assistant.render(
                            {
                                "thought": iteration.state.thought,
                                "tool_name": iteration.state.tool_name,
                                "tool_input": iteration.state.tool_input,
                                "tool_output": tool_result.output.to_string(),
                                "final_answer": iteration.state.final_answer,
                            }
                        ),
                        meta=MessageMeta({"success": tool_result.success}),
                    )
                )
                iteration.state.tool_output = tool_result.output.get_text_content()

                for key in ["partialUpdate", "update"]:
                    await iteration.emitter.emit(
                        key,
                        {
                            "data": iteration.state,
                            "update": {
                                "key": "tool_output",
                                "value": tool_result.output,
                                "parsedValue": tool_result.output.to_string(),
                            },
                            "meta": {"success": tool_result.success},  # TODO deleted meta
                            "memory": runner.memory,
                        },
                    )

            if iteration.state.final_answer:
                final_message = AssistantMessage(
                    content=iteration.state.final_answer, meta=MessageMeta({"createdAt": datetime.now(tz=UTC)})
                )
                await runner.memory.add(final_message)
                await iteration.emitter.emit(
                    "success",
                    {
                        "data": final_message,
                        "iterations": runner.iterations,
                        "memory": runner.memory,
                        "meta": iteration.meta,
                    },
                )

        if run_input.prompt is not None:
            await self.input.memory.add(
                UserMessage(content=run_input.prompt, meta=MessageMeta({"createdAt": context.created_at}))
            )

        await self.input.memory.add(final_message)

        return BeeRunOutput(result=final_message, iterations=runner.iterations, memory=runner.memory)
